tests/test_read_data.py

```diff
@@ -0,0 +1,397 @@
+    file_names = df["filename"].unique().compute()
+    if backend == "cudf":
+        file_names = file_names.to_pandas()
+
+    assert len(file_names) == NUM_FILES
+    assert set(file_names.values) == {
+        f"test_{file_id}.{file_type}" for file_id in range(NUM_FILES)
+    }
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+@pytest.mark.parametrize(
+    "file_type,add_filename,function_name",
+    [
+        *[("jsonl", True, func) for func in ["read_data_blocksize", "read_data_fpp"]],
+        *[("jsonl", False, func) for func in ["read_data_blocksize", "read_data_fpp"]],
+        *[
+            ("parquet", False, func)
+            for func in ["read_data_blocksize", "read_data_fpp"]
+        ],
+        *[("parquet", True, "read_data_fpp")],
+    ],
+)
+@pytest.mark.parametrize(
+    "cols_to_select", [None, ["id"], ["text", "id"], ["id", "text"]]
+)
+def test_read_data_select_columns(
+    mock_multiple_jsonl_files,
+    mock_multiple_parquet_files,
+    backend,
+    file_type,
+    add_filename,
+    function_name,
+    cols_to_select,
+):
+    input_files = (
+        mock_multiple_jsonl_files
+        if file_type == "jsonl"
+        else mock_multiple_parquet_files
+    )
+    if function_name == "read_data_fpp":
+        func = read_data_fpp
+        read_kwargs = {"files_per_partition": 1}
+    elif function_name == "read_data_blocksize":
+        func = read_data_blocksize
+        read_kwargs = {"blocksize": "128Mib"}
+
+    df = func(
+        input_files=input_files,
+        backend=backend,
+        file_type=file_type,
+        add_filename=add_filename,
+        input_meta=None,
+        columns=list(cols_to_select) if cols_to_select else None,
+        **read_kwargs,
+    )
+    if not cols_to_select:
+        cols_to_select = ["id", "text"]
+
+    if not add_filename:
+        assert list(df.columns) == sorted(cols_to_select)
+    else:
+        assert list(df.columns) == sorted(cols_to_select + ["filename"])
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+@pytest.mark.parametrize("function_name", ["read_data_blocksize", "read_data_fpp"])
+@pytest.mark.parametrize(
+    "input_meta", [{"id": "str"}, {"text": "str"}, {"id": "str", "text": "str"}]
+)
+def test_read_data_input_data(
+    mock_multiple_jsonl_files, backend, function_name, input_meta
+):
+    if function_name == "read_data_fpp":
+        func = read_data_fpp
+        read_kwargs = {"files_per_partition": 1}
+    elif function_name == "read_data_blocksize":
+        func = read_data_blocksize
+        read_kwargs = {"blocksize": "128Mib"}
+
+    df = func(
+        input_files=mock_multiple_jsonl_files,
+        backend=backend,
+        file_type="jsonl",
+        add_filename=False,
+        input_meta=input_meta,
+        columns=None,
+        **read_kwargs,
+    )
+
```

Comment:
Please TAL here as the behavior of when we pass `input_meta` results in different outputs depending on `fpp` vs `blocksize`

Reply:
Okay, I think I understand what is being demonstrated here.

Even though Pandas `read_data_fpp` and Pandas/cuDF `read_data_blocksize` cannot prune columns while reading, can't we still have it prune the columns after reading? I know that doesn't save on I/O but then at least the user is getting the resulting DataFrame that they would expect.

Reply:
That's a good point. Would you want to do a `select_columns([input_meta.keys()])` when input_type=='jsonl' (since input_meta in it's current setting only supports `json`)? We can do that, though I worry that it'll become convolved along with `columns` (and maybe even `add_filename`)

Reply:
I think as long as we are thoroughly checking all the conditions we expect to have to do this, it should be okay. Happy to discuss this offline if anything is unclear here.

Reply:
@sarahyurick I made the changes in the last commit however the `test_io` now fails due to since the test does

```python
            dataset = DocumentDataset.read_json(
                temp_file.name, input_meta='{"id": "float"}'
            )

        output_meta = str({col: str(dtype) for col, dtype in dataset.df.dtypes.items()})

        expected_meta = (
            "{'date': 'datetime64[ns, UTC]', 'id': 'float64', 'text': 'object'}"
        )
```

I was going to change it but just wanted to confirm if it's okay. FWIW, that test only passes in pandas backend case, but would've failed if backend would've been cudf.